File: src/lib/tripSettings.ts
import type { CSSProperties } from 'react';

export interface TripCopyLabels {
  itineraryTab: string;
  mapsTab: string;
  draftTab: string;
  budgetTab: string;
  checklistTab: string;
  documentsTab: string;
  photosTab: string;
  searchPlaceholder: string;
  overviewEyebrow: string;
  overviewIntroFilled: string;
  overviewIntroEmpty: string;
  backToOverview: string;
  customizePlan: string;
  doneCustomizing: string;
  resetPlan: string;
  photosButton: string;
  dayLabel: string;
  daysLabel: string;
  currentLocationLabel: string;
  spotsSuffix: string;
  openMapLabel: string;
  activityPhotosLabel: string;
  deleteActivityLabel: string;
  deleteActivityConfirm: string;
}

export interface TripThemeSettings {
  bg: string;
  bgElevated: string;
  ink: string;
  inkMuted: string;
  accent: string;
  accentSoft: string;
}

export type ThemeMode = 'light' | 'dark';

export interface ThemePalettePreset {
  id: string;
  name: string;
  description: string;
  light: TripThemeSettings;
  dark: TripThemeSettings;
}

export const DEFAULT_LIGHT_THEME: TripThemeSettings = {
  bg: '#FAF7F2',
  bgElevated: '#FFFFFF',
  ink: '#0F0E0D',
  inkMuted: '#5C5853',
  /* The journey green is the app's own accent. A trip that has not chosen a
     palette inherits it, so nothing renders in the legacy pink by default. */
  accent: '#174b38',
  accentSoft: '#e4eadc',
};

export const DEFAULT_DARK_THEME: TripThemeSettings = {
  bg: '#14110F',
  bgElevated: '#1F1A17',
  ink: '#F5EFE4',
  inkMuted: '#A39B8C',
  accent: '#2f7559',
  accentSoft: '#263c31',
};

/** Theme families with paired light + dark variants. One tap applies both. */
export const THEME_PALETTE_PRESETS: ThemePalettePreset[] = [
  {
    id: 'ember-rose',
    name: 'Ember Rose',
    description: 'Warm paper by day, charcoal rose by night.',
    light: { ...DEFAULT_LIGHT_THEME },
    dark: { ...DEFAULT_DARK_THEME },
  },
  {
    id: 'quantum-rose',
    name: 'Quantum Rose',
    description: 'A vivid pink interface with lilac depth and editorial contrast.',
    light: {
      bg: '#FFF0F8',
      bgElevated: '#FFF7FC',
      ink: '#91185C',
      inkMuted: '#A82A70',
      accent: '#E6067A',
      accentSoft: '#FFD6FF',
    },
    dark: {
      bg: '#1A0922',
      bgElevated: '#2A1435',
      ink: '#FFB3FF',
      inkMuted: '#D67AD6',
      accent: '#FF6BEF',
      accentSoft: '#46204F',
    },
  },
  {
    id: 'bubblegum',
    name: 'Bubblegum',
    description: 'Playful pink energy balanced by citrus cream and pool-blue calm.',
    light: {
      bg: '#F6E6EE',
      bgElevated: '#FDEDC9',
      ink: '#5B5B5B',
      inkMuted: '#7A7A7A',
      accent: '#D04F99',
      accentSoft: '#8ACFD1',
    },
    dark: {
      bg: '#12242E',
      bgElevated: '#1C2E38',
      ink: '#F3E3EA',
      inkMuted: '#E4A2B1',
      accent: '#FBE2A7',
      accentSoft: '#E4A2B1',
    },
  },
  {
    id: 'lavender',
    name: 'Lavender',
    description: 'Quiet violet surfaces with soft rose warmth and steady contrast.',
    light: {
      bg: '#F8F7FA',
      bgElevated: '#FFFFFF',
      ink: '#3D3C4F',
      inkMuted: '#6B6880',
      accent: '#8A79AB',
      accentSoft: '#DFD9EC',
    },
    dark: {
      bg: '#1A1823',
      bgElevated: '#232030',
      ink: '#E0DDEF',
      inkMuted: '#A09AAD',
      accent: '#A995C9',
      accentSoft: '#5A5370',
    },
  },
  {
    id: 'primary-pop',
    name: 'Primary Pop',
    description: 'High-contrast primary colors for a direct, graphic interface.',
    light: {
      bg: '#FFFFFF',
      bgElevated: '#FFFFFF',
      ink: '#000000',
      inkMuted: '#333333',
      accent: '#FF3333',
      accentSoft: '#FFFF00',
    },
    dark: {
      bg: '#000000',
      bgElevated: '#333333',
      ink: '#FFFFFF',
      inkMuted: '#CCCCCC',
      accent: '#FF6666',
      accentSoft: '#FFFF33',
    },
  },
  {
    id: 'garden-indigo',
    name: 'Garden Indigo',
    description: 'Leafy greens with indigo structure, teal lift, and a warm amber signal.',
    light: {
      bg: '#F7F9F3',
      bgElevated: '#FFFFFF',
      ink: '#000000',
      inkMuted: '#333333',
      accent: '#4F46E5',
      accentSoft: '#14B8A6',
    },
    dark: {
      bg: '#000000',
      bgElevated: '#1A212B',
      ink: '#FFFFFF',
      inkMuted: '#CCCCCC',
      accent: '#818CF8',
      accentSoft: '#2DD4BF',
    },
  },
  {
    id: 'terracotta-coast',
    name: 'Terracotta Coast',
    description: 'Weathered stone, terracotta warmth, and a restrained coastal blue.',
    light: {
      bg: '#E8EBED',
      bgElevated: '#FFFFFF',
      ink: '#333333',
      inkMuted: '#6B7280',
      accent: '#E05D38',
      accentSoft: '#D6E4F0',
    },
    dark: {
      bg: '#1C2433',
      bgElevated: '#2A3040',
      ink: '#E5E5E5',
      inkMuted: '#A3A3A3',
      accent: '#E05D38',
      accentSoft: '#2A3656',
    },
  },
];

export const themesMatch = (a: TripThemeSettings, b: TripThemeSettings) =>
  (Object.keys(a) as Array<keyof TripThemeSettings>).every(
    (key) => a[key].toUpperCase() === b[key].toUpperCase(),
  );

export const getPresetVariant = (preset: ThemePalettePreset, mode: ThemeMode) =>
  mode === 'light' ? preset.light : preset.dark;

export const findMatchingThemePreset = (
  settings: Pick<TripAppSettings, 'theme' | 'lightTheme'>,
  mode?: ThemeMode,
) => {
  if (mode) {
    const current = mode === 'light' ? settings.lightTheme : settings.theme;
    return (
      THEME_PALETTE_PRESETS.find((preset) => themesMatch(getPresetVariant(preset, mode), current)) ?? null
    );
  }

  return (
    THEME_PALETTE_PRESETS.find(
      (preset) =>
        themesMatch(preset.light, settings.lightTheme) && themesMatch(preset.dark, settings.theme),
    ) ?? null
  );
};

export const getThemeForMode = (settings: Pick<TripAppSettings, 'theme' | 'lightTheme'>, mode: ThemeMode) =>
  mode === 'light' ? settings.lightTheme : settings.theme;

export const buildTripThemeStyle = (palette: TripThemeSettings, mode: ThemeMode): CSSProperties => {
  const border = `color-mix(in srgb, ${palette.ink} 22%, ${palette.bgElevated})`;
  const sharedTokens = {
    '--background': palette.bg,
    '--foreground': palette.ink,
    '--card': palette.bgElevated,
    '--card-foreground': palette.ink,
    '--popover': palette.bgElevated,
    '--popover-foreground': palette.ink,
    '--primary': palette.accent,
    '--primary-foreground': '#0F0E0D',
    '--secondary': palette.accentSoft,
    '--secondary-foreground': palette.ink,
    '--muted': palette.accentSoft,
    '--muted-foreground': palette.inkMuted,
    '--accent-foreground': '#0F0E0D',
    '--input': border,
    '--ring': palette.accent,
  };

  if (mode === 'light') {
    return {
      ...sharedTokens,
      '--bg': palette.bg,
      '--bg-elevated': palette.bgElevated,
      '--ink': palette.ink,
      '--ink-muted': palette.inkMuted,
      '--accent': palette.accent,
      '--accent-soft': palette.accentSoft,
      '--accent-ink': '#0F0E0D',
      '--border': border,
      '--shadow-lift': '0 1px 0 rgba(15,14,13,0.04), 0 12px 32px -16px rgba(15,14,13,0.18)',
    } as CSSProperties;
  }

  return {
    ...sharedTokens,
    '--bg': palette.bg,
    '--bg-elevated': palette.bgElevated,
    '--ink': palette.ink,
    '--ink-muted': palette.inkMuted,
    '--accent': palette.accent,
    '--accent-soft': palette.accentSoft,
    '--accent-ink': '#0F0E0D',
    '--border': border,
    '--shadow-lift': '0 1px 0 rgba(0,0,0,0.3), 0 18px 40px -18px rgba(0,0,0,0.6)',
  } as CSSProperties;
};

export interface TripAppSettings {
  heroEyebrow: string;
  heroHeadline: string;
  heroDescription: string;
  heroPrimaryCta: string;
  heroSecondaryCta: string;
  coverLabel: string;
  coverHeadline: string;
  coverStatusEmpty: string;
  coverStatusFilled: string;
  coverModeEmpty: string;
  coverModeFilled: string;
  marqueeItems: string[];
  coverImage: string | null;
  immersiveEffects: boolean;
  labels: TripCopyLabels;
  /** Dark-mode palette (legacy `theme` field). */
  theme: TripThemeSettings;
  /** Light-mode palette. */
  lightTheme: TripThemeSettings;
  /** Optional user-saved palette shown alongside the built-in presets. */
  customThemePreset: ThemePalettePreset | null;
}

export const DEFAULT_TRIP_SETTINGS: TripAppSettings = {
  heroEyebrow: 'A personalized travel starter',
  heroHeadline: 'Plan your next trip your way.',
  heroDescription: 'Add cities, days, notes, budgets, maps, and documents as you build your travel plan.',
  heroPrimaryCta: 'Open handbook',
  heroSecondaryCta: 'Open maps',
  coverLabel: 'Custom cover',
  coverHeadline: 'Add your\nown story',
  coverStatusEmpty: 'No cities yet',
  coverStatusFilled: '{cities}',
  coverModeEmpty: 'starter',
  coverModeFilled: 'handbook',
  marqueeItems: ['Travel Handbook', 'Plans', 'Notes', 'Maps', 'Photos'],
  coverImage: null,
  immersiveEffects: false,
  labels: {
    itineraryTab: 'Itinerary',
    mapsTab: 'Maps',
    draftTab: 'Draft',
    budgetTab: 'Budget',
    checklistTab: 'Checklist',
    documentsTab: 'Documents',
    photosTab: 'Photo Wall',
    searchPlaceholder: 'Search itinerary or locations...',
    overviewEyebrow: 'The itinerary · day by day',
    overviewIntroFilled: 'A day-by-day field guide for {cities}.',
    overviewIntroEmpty: 'A blank day-by-day field guide ready for your trip details.',
    backToOverview: 'Back to Overview',
    customizePlan: 'Customize Plan',
    doneCustomizing: 'Done Customizing',
    resetPlan: 'Reset',
    photosButton: 'Photos',
    dayLabel: 'Day',
    daysLabel: 'days',
    currentLocationLabel: 'Current Location',
    spotsSuffix: 'spots',
    openMapLabel: 'Map',
    activityPhotosLabel: 'Photos',
    deleteActivityLabel: 'Delete',
    deleteActivityConfirm: 'Delete this activity?',
  },
  theme: { ...DEFAULT_DARK_THEME },
  lightTheme: { ...DEFAULT_LIGHT_THEME },
  customThemePreset: null,
};

export const mergeTripSettings = (settings?: Partial<TripAppSettings> | null): TripAppSettings => ({
  ...DEFAULT_TRIP_SETTINGS,
  ...settings,
  marqueeItems:
    settings?.marqueeItems && settings.marqueeItems.length > 0
      ? settings.marqueeItems
      : DEFAULT_TRIP_SETTINGS.marqueeItems,
  labels: {
    ...DEFAULT_TRIP_SETTINGS.labels,
    ...(settings?.labels || {}),
  },
  theme: {
    ...DEFAULT_TRIP_SETTINGS.theme,
    ...(settings?.theme || {}),
  },
  lightTheme: {
    ...DEFAULT_TRIP_SETTINGS.lightTheme,
    ...(settings?.lightTheme || {}),
  },
  customThemePreset: settings?.customThemePreset
    ? {
        ...settings.customThemePreset,
        light: { ...settings.customThemePreset.light },
        dark: { ...settings.customThemePreset.dark },
      }
    : null,
});

export const applyTemplate = (template: string, replacements: Record<string, string>) =>
  template.replace(/\{(\w+)\}/g, (_, key: string) => replacements[key] ?? '');

File: TravelHandbook-iOS/TravelHandbook/Models/TripSettings.swift
import Foundation

struct TripCopyLabels: Codable, Equatable {
    var itineraryTab: String
    var mapsTab: String
    var draftTab: String
    var budgetTab: String
    var checklistTab: String
    var documentsTab: String
    var photosTab: String
    var searchPlaceholder: String
    var overviewEyebrow: String
    var overviewIntroFilled: String
    var overviewIntroEmpty: String
    var backToOverview: String
    var customizePlan: String
    var doneCustomizing: String
    var resetPlan: String
    var photosButton: String
    var dayLabel: String
    var daysLabel: String
    var currentLocationLabel: String
    var spotsSuffix: String
    var openMapLabel: String
    var activityPhotosLabel: String
    var deleteActivityLabel: String
    var deleteActivityConfirm: String
}

struct TripThemeSettings: Codable, Equatable {
    var bg: String
    var bgElevated: String
    var ink: String
    var inkMuted: String
    var accent: String
    var accentSoft: String
}

enum ThemeMode: String, Codable, CaseIterable {
    case light, dark
}

struct ThemePalettePreset: Codable, Equatable, Identifiable {
    var id: String
    var name: String
    var description: String
    var light: TripThemeSettings
    var dark: TripThemeSettings
}

let DEFAULT_LIGHT_THEME = TripThemeSettings(
    bg: "#FAF7F2",
    bgElevated: "#FFFFFF",
    ink: "#0F0E0D",
    inkMuted: "#5C5853",
    accent: "#EE4D87",
    accentSoft: "#FFE4EE"
)

let DEFAULT_DARK_THEME = TripThemeSettings(
    bg: "#14110F",
    bgElevated: "#1F1A17",
    ink: "#F5EFE4",
    inkMuted: "#A39B8C",
    accent: "#FF6B9A",
    accentSoft: "#3A1F2A"
)

let THEME_PALETTE_PRESETS: [ThemePalettePreset] = [
    ThemePalettePreset(
        id: "ember-rose",
        name: "Ember Rose",
        description: "Warm paper by day, charcoal rose by night.",
        light: DEFAULT_LIGHT_THEME,
        dark: DEFAULT_DARK_THEME
    ),
    ThemePalettePreset(
        id: "midnight-slate",
        name: "Midnight Slate",
        description: "Cool mist in light mode, slate ink after dark.",
        light: TripThemeSettings(
            bg: "#F4F5F8",
            bgElevated: "#FFFFFF",
            ink: "#171A22",
            inkMuted: "#667085",
            accent: "#E7685D",
            accentSoft: "#F7D6DD"
        ),
        dark: TripThemeSettings(
            bg: "#171A22",
            bgElevated: "#232630",
            ink: "#F7F2EB",
            inkMuted: "#8E8678",
            accent: "#E7685D",
            accentSoft: "#41242C"
        )
    ),
    ThemePalettePreset(
        id: "forest",
        name: "Forest",
        description: "Linen green calm that deepens at night.",
        light: TripThemeSettings(
            bg: "#F3F5F4",
            bgElevated: "#FFFFFF",
            ink: "#15201C",
            inkMuted: "#5F6B66",
            accent: "#2F7D6E",
            accentSoft: "#DDEFEA"
        ),
        dark: TripThemeSettings(
            bg: "#121714",
            bgElevated: "#1B221D",
            ink: "#EEF3EC",
            inkMuted: "#8FA093",
            accent: "#2F7D6E",
            accentSoft: "#1E332C"
        )
    ),
    ThemePalettePreset(
        id: "coastal",
        name: "Coastal",
        description: "Sky-sheet pages with deep navy evenings.",
        light: TripThemeSettings(
            bg: "#F2F6F9",
            bgElevated: "#FFFFFF",
            ink: "#132033",
            inkMuted: "#5C6B7A",
            accent: "#3D8FB5",
            accentSoft: "#D7EAF3"
        ),
        dark: TripThemeSettings(
            bg: "#10161C",
            bgElevated: "#182129",
            ink: "#EEF4F7",
            inkMuted: "#8A9AA6",
            accent: "#3D8FB5",
            accentSoft: "#1A303A"
        )
    ),
    ThemePalettePreset(
        id: "amber",
        name: "Amber",
        description: "Sunlit sandstone that turns lamp-warm at night.",
        light: TripThemeSettings(
            bg: "#F7F0E8",
            bgElevated: "#FFFBF6",
            ink: "#1A140E",
            inkMuted: "#6B655D",
            accent: "#C8842A",
            accentSoft: "#F3E2C8"
        ),
        dark: TripThemeSettings(
            bg: "#16120C",
            bgElevated: "#221B13",
            ink: "#F8F0E3",
            inkMuted: "#A89880",
            accent: "#E0A045",
            accentSoft: "#3A2C18"
        )
    )
]

struct TripAppSettings: Codable, Equatable {
    var heroEyebrow: String
    var heroHeadline: String
    var heroDescription: String
    var heroPrimaryCta: String
    var heroSecondaryCta: String
    var coverLabel: String
    var coverHeadline: String
    var coverStatusEmpty: String
    var coverStatusFilled: String
    var coverModeEmpty: String
    var coverModeFilled: String
    var marqueeItems: [String]
    var coverImage: String?
    var immersiveEffects: Bool
    var labels: TripCopyLabels
    var theme: TripThemeSettings
    var lightTheme: TripThemeSettings
}

let DEFAULT_TRIP_SETTINGS = TripAppSettings(
    heroEyebrow: "A personalized travel starter",
    heroHeadline: "Plan your next trip your way.",
    heroDescription: "Add cities, days, notes, budgets, maps, and documents as you build your travel plan.",
    heroPrimaryCta: "Open handbook",
    heroSecondaryCta: "Open maps",
    coverLabel: "Custom cover",
    coverHeadline: "Add your\nown story",
    coverStatusEmpty: "No cities yet",
    coverStatusFilled: "{cities}",
    coverModeEmpty: "starter",
    coverModeFilled: "handbook",
    marqueeItems: ["Travel Handbook", "Plans", "Notes", "Maps", "Photos"],
    coverImage: nil,
    immersiveEffects: false,
    labels: TripCopyLabels(
        itineraryTab: "Itinerary",
        mapsTab: "Maps",
        draftTab: "Draft",
        budgetTab: "Budget",
        checklistTab: "Checklist",
        documentsTab: "Documents",
        photosTab: "Photo Wall",
        searchPlaceholder: "Search itinerary or locations...",
        overviewEyebrow: "The itinerary · day by day",
        overviewIntroFilled: "A day-by-day field guide for {cities}.",
        overviewIntroEmpty: "A blank day-by-day field guide ready for your trip details.",
        backToOverview: "Back to Overview",
        customizePlan: "Customize Plan",
        doneCustomizing: "Done Customizing",
        resetPlan: "Reset",
        photosButton: "Photos",
        dayLabel: "Day",
        daysLabel: "days",
        currentLocationLabel: "Current Location",
        spotsSuffix: "spots",
        openMapLabel: "Map",
        activityPhotosLabel: "Photos",
        deleteActivityLabel: "Delete",
        deleteActivityConfirm: "Delete this activity?"
    ),
    theme: DEFAULT_DARK_THEME,
    lightTheme: DEFAULT_LIGHT_THEME
)

func mergeThemeSettings(_ base: TripThemeSettings, _ override: TripThemeSettings) -> TripThemeSettings {
    TripThemeSettings(
        bg: override.bg.isEmpty ? base.bg : override.bg,
        bgElevated: override.bgElevated.isEmpty ? base.bgElevated : override.bgElevated,
        ink: override.ink.isEmpty ? base.ink : override.ink,
        inkMuted: override.inkMuted.isEmpty ? base.inkMuted : override.inkMuted,
        accent: override.accent.isEmpty ? base.accent : override.accent,
        accentSoft: override.accentSoft.isEmpty ? base.accentSoft : override.accentSoft
    )
}

func mergeCopyLabels(_ base: TripCopyLabels, _ override: TripCopyLabels) -> TripCopyLabels {
    TripCopyLabels(
        itineraryTab: override.itineraryTab.isEmpty ? base.itineraryTab : override.itineraryTab,
        mapsTab: override.mapsTab.isEmpty ? base.mapsTab : override.mapsTab,
        draftTab: override.draftTab.isEmpty ? base.draftTab : override.draftTab,
        budgetTab: override.budgetTab.isEmpty ? base.budgetTab : override.budgetTab,
        checklistTab: override.checklistTab.isEmpty ? base.checklistTab : override.checklistTab,
        documentsTab: override.documentsTab.isEmpty ? base.documentsTab : override.documentsTab,
        photosTab: override.photosTab.isEmpty ? base.photosTab : override.photosTab,
        searchPlaceholder: override.searchPlaceholder.isEmpty ? base.searchPlaceholder : override.searchPlaceholder,
        overviewEyebrow: override.overviewEyebrow.isEmpty ? base.overviewEyebrow : override.overviewEyebrow,
        overviewIntroFilled: override.overviewIntroFilled.isEmpty ? base.overviewIntroFilled : override.overviewIntroFilled,
        overviewIntroEmpty: override.overviewIntroEmpty.isEmpty ? base.overviewIntroEmpty : override.overviewIntroEmpty,
        backToOverview: override.backToOverview.isEmpty ? base.backToOverview : override.backToOverview,
        customizePlan: override.customizePlan.isEmpty ? base.customizePlan : override.customizePlan,
        doneCustomizing: override.doneCustomizing.isEmpty ? base.doneCustomizing : override.doneCustomizing,
        resetPlan: override.resetPlan.isEmpty ? base.resetPlan : override.resetPlan,
        photosButton: override.photosButton.isEmpty ? base.photosButton : override.photosButton,
        dayLabel: override.dayLabel.isEmpty ? base.dayLabel : override.dayLabel,
        daysLabel: override.daysLabel.isEmpty ? base.daysLabel : override.daysLabel,
        currentLocationLabel: override.currentLocationLabel.isEmpty ? base.currentLocationLabel : override.currentLocationLabel,
        spotsSuffix: override.spotsSuffix.isEmpty ? base.spotsSuffix : override.spotsSuffix,
        openMapLabel: override.openMapLabel.isEmpty ? base.openMapLabel : override.openMapLabel,
        activityPhotosLabel: override.activityPhotosLabel.isEmpty ? base.activityPhotosLabel : override.activityPhotosLabel,
        deleteActivityLabel: override.deleteActivityLabel.isEmpty ? base.deleteActivityLabel : override.deleteActivityLabel,
        deleteActivityConfirm: override.deleteActivityConfirm.isEmpty ? base.deleteActivityConfirm : override.deleteActivityConfirm
    )
}

func mergeTripSettings(_ settings: TripAppSettings?) -> TripAppSettings {
    guard let settings else { return DEFAULT_TRIP_SETTINGS }

    var merged = DEFAULT_TRIP_SETTINGS
    merged.heroEyebrow = settings.heroEyebrow
    merged.heroHeadline = settings.heroHeadline
    merged.heroDescription = settings.heroDescription
    merged.heroPrimaryCta = settings.heroPrimaryCta
    merged.heroSecondaryCta = settings.heroSecondaryCta
    merged.coverLabel = settings.coverLabel
    merged.coverHeadline = settings.coverHeadline
    merged.coverStatusEmpty = settings.coverStatusEmpty
    merged.coverStatusFilled = settings.coverStatusFilled
    merged.coverModeEmpty = settings.coverModeEmpty
    merged.coverModeFilled = settings.coverModeFilled
    merged.coverImage = settings.coverImage
    merged.immersiveEffects = settings.immersiveEffects
    merged.marqueeItems = settings.marqueeItems.isEmpty ? DEFAULT_TRIP_SETTINGS.marqueeItems : settings.marqueeItems
    merged.labels = mergeCopyLabels(DEFAULT_TRIP_SETTINGS.labels, settings.labels)
    merged.theme = mergeThemeSettings(DEFAULT_TRIP_SETTINGS.theme, settings.theme)
    merged.lightTheme = mergeThemeSettings(DEFAULT_TRIP_SETTINGS.lightTheme, settings.lightTheme)
    return merged
}

struct ShellThemePalettes: Codable, Equatable {
    var light: TripThemeSettings
    var dark: TripThemeSettings

    static let `default` = ShellThemePalettes(light: DEFAULT_LIGHT_THEME, dark: DEFAULT_DARK_THEME)
}

func themesMatch(_ a: TripThemeSettings, _ b: TripThemeSettings) -> Bool {
    a.bg.caseInsensitiveCompare(b.bg) == .orderedSame
        && a.bgElevated.caseInsensitiveCompare(b.bgElevated) == .orderedSame
        && a.ink.caseInsensitiveCompare(b.ink) == .orderedSame
        && a.inkMuted.caseInsensitiveCompare(b.inkMuted) == .orderedSame
        && a.accent.caseInsensitiveCompare(b.accent) == .orderedSame
        && a.accentSoft.caseInsensitiveCompare(b.accentSoft) == .orderedSame
}

func getPresetVariant(_ preset: ThemePalettePreset, mode: ThemeMode) -> TripThemeSettings {
    mode == .light ? preset.light : preset.dark
}

func getThemeForMode(settings: TripAppSettings, mode: ThemeMode) -> TripThemeSettings {
    mode == .light ? settings.lightTheme : settings.theme
}

func applyTemplate(_ template: String, replacements: [String: String]) -> String {
    var result = template
    for (key, value) in replacements {
        result = result.replacingOccurrences(of: "{\(key)}", with: value)
    }
    return result
}

func shellThemeFromTripSettings(_ settings: TripAppSettings) -> ShellThemePalettes {
    let merged = mergeTripSettings(settings)
    return ShellThemePalettes(light: merged.lightTheme, dark: merged.theme)
}
